core/chains/evm/chain_set.go

```diff
@@ -268,6 +274,75 @@ func (cll *chainSet) ORM() types.ORM {
 	return cll.orm
 }
 
+func (cll *chainSet) GetNode(ctx context.Context, id int32) (n evmtypes.Node, err error) {
+	n, err = cll.orm.Node(id, pg.WithParentCtx(ctx))
+	if err != nil {
+		err = errors.Wrap(err, "GetNode failed to load node from DB")
+		return
+	}
+	cll.addStateToNode(&n)
+	return
+}
+
+func (cll *chainSet) GetNodes(ctx context.Context, offset, limit int) (nodes []evmtypes.Node, count int, err error) {
+	nodes, count, err = cll.orm.Nodes(offset, limit, pg.WithParentCtx(ctx))
+	if err != nil {
+		err = errors.Wrap(err, "GetNodes failed to load nodes from DB")
+		return
+	}
+	for i := range nodes {
+		cll.addStateToNode(&nodes[i])
+	}
+	return
+}
+
+func (cll *chainSet) GetNodesForChain(ctx context.Context, chainID utils.Big, offset, limit int) (nodes []evmtypes.Node, count int, err error) {
+	nodes, count, err = cll.orm.NodesForChain(chainID, offset, limit, pg.WithParentCtx(ctx))
+	if err != nil {
+		err = errors.Wrap(err, "GetNodesForChain failed to load nodes from DB")
+		return
+	}
+	for i := range nodes {
+		cll.addStateToNode(&nodes[i])
+	}
+	return
+}
+
+func (cll *chainSet) GetNodesByChainIDs(ctx context.Context, chainIDs []utils.Big) (nodes []evmtypes.Node, err error) {
+	nodes, err = cll.orm.GetNodesByChainIDs(chainIDs, pg.WithParentCtx(ctx))
+	if err != nil {
+		err = errors.Wrap(err, "GetNodesForChain failed to load nodes from DB")
+		return
+	}
+	for i := range nodes {
+		cll.addStateToNode(&nodes[i])
+	}
+	return
+}
+
+func (cll *chainSet) addStateToNode(n *evmtypes.Node) {
```

Comment:
I have made efforts to ensure that querying `node.State()` should always be relatively fast but I suppose its possible it could block for a "long" time.

I am not sure either what the best approach is.